Add lxd connection plugin.

mattclay opened the pull request:
##### Issue Type:

Feature Pull Request
##### Ansible Version:

```
ansible 2.1.0 (lxd-plugin 43929197f1) last updated 2016/03/04 13:37:19 (GMT -700)
  lib/ansible/modules/core: (detached HEAD 0bbb7ba38d) last updated 2016/03/03 15:34:52 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD 39e4040685) last updated 2016/03/03 15:34:52 (GMT -700)
  config file = 
  configured module search path = Default w/o overrides
```
##### Summary:

Add a connection plugin for [lxd](https://linuxcontainers.org/lxd/). Connects to a local or remote lxd server using the lxc command line client. Works with and without pipelining.

Tested on a Ubuntu 15.10 control host with a Ubuntu 15.10 lxd host, using the official lxd install as well as the official PPA version. Tested manually as well as with the tests in PR ansible/ansible#14812.
##### Example output:

Connecting to a local container:

```
$ ansible -m ping -i 'centos-7-amd64,' all -e 'ansible_connection=lxd'
centos-7-amd64 | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
```

Connecting to a remote container:

```
$ ansible -m ping -i 'iron:ubuntu-wily-amd64-os,' all -e 'ansible_connection=lxd'
iron:ubuntu-wily-amd64-os | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
```

If a container is not running:

```
$ ansible -m ping -i 'iron:ubuntu-wily-amd64-pip,' all -e 'ansible_connection=lxd'
iron:ubuntu-wily-amd64-pip | FAILED! => {
    "failed": true, 
    "msg": "container not running: iron:ubuntu-wily-amd64-pip"
}
```

If a container is not found:

```
$ ansible -m ping -i 'iron:does-not-exist,' all -e 'ansible_connection=lxd'
iron:does-not-exist | FAILED! => {
    "failed": true, 
    "msg": "container not found: iron:does-not-exist"
}
```

If the lxc command line client is not found:

```
# ansible -m ping -i 'iron:does-not-exist,' all -e 'ansible_connection=lxd'
iron:does-not-exist | FAILED! => {
    "failed": true, 
    "msg": "lxc command not found in PATH"
}
```

mattclay committed:
Add lxd connection plugin.

bcoca commented on lib/ansible/plugins/connection/lxd.py:
```diff
@@ -0,0 +1,116 @@
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
```
you should not need this as connectionbase already has self._display

mattclay replied:
Thanks for pointing out the extra display. I've fixed that.

bcoca commented on lib/ansible/plugins/connection/lxd.py:
```diff
@@ -0,0 +1,116 @@
+
+        display.vvv(u"EXEC {0}".format(cmd), host=self._host)
+
+        executable = C.DEFAULT_EXECUTABLE.split()[0] if C.DEFAULT_EXECUTABLE else "/bin/sh"
```
action plugin should have already taken care of this

nitzmahone replied:
Yes and no- the action plugin sticks the right executable wrapper on there, but it returns it as a single string (which absolutely won't work when passed directly to lxc- I checked). This second level of wrap lets the shell on the remote-side parse that command- the alternative would be shlex.split(), but that would make the behavior of "raw" inconsistent in this plugin compared to all others. eg:
```
raw: echo hi && echo there
```
would be run under a shell and return "hi there" on every other connection plugin type, but for this one would return "hi && echo there" because a shell didn't parse it first. 

Seems like we have to do the remote shell if we want raw's behavior to remain consistent...

bcoca commented on lib/ansible/plugins/connection/lxd.py:
```diff
@@ -0,0 +1,116 @@
+        local_cmd = [self._lxc_cmd, "exec", self._host, "--", executable, "-c", cmd]
+
+        local_cmd = map(to_bytes, local_cmd)
+        in_data = to_bytes(in_data)
```
Not sure you want to do this if in_data is None

mattclay replied:
I've fixed the to_bytes call on in_data to return None if that's what's given. This is the same as what the ssh plugin does.

nitzmahone commented on lib/ansible/plugins/connection/lxd.py:
```diff
@@ -0,0 +1,110 @@
+
+        self._display.vvv(u"EXEC {0}".format(cmd), host=self._host)
+
+        executable = C.DEFAULT_EXECUTABLE.split()[0] if C.DEFAULT_EXECUTABLE else "/bin/sh"
```
@bcoca - your comment about the action_plugin already doing this is half-correct- the action plugin sticks the right executable wrapper on there, but it returns the command as a single string (which absolutely won't work when passed directly to lxc- I checked). This second level of wrap lets the shell on the remote-side parse that command- the alternative would be shlex.split(), but that would make the behavior of "raw" inconsistent in this plugin compared to all others. eg:

```
raw: echo hi && echo there
```

would be run under a shell and return "hi there" on every other connection plugin type, but for this one would return "hi && echo there" because a shell didn't parse it first.

Seems like we have to do the second remote shell if we want raw's behavior to remain consistent...

bcoca replied:
we can probably change it to a list, still this would be the wrong executable in some cases as it does not allow per host config, its also redundant to fallback to /bin/sh as it always is set (default is /bin/sh already).

In other cases it will still be a problem, specially when privilege escalation is involved, due to the quoting.

mattclay replied:
Looking over all the connection plugins, I'm not sure changing cmd to a list actually helps the situation. Every connection plugin passes cmd through a shell, either explicitly within the plugin or by way of delegating that responsibility to the remote host.

These plugins use subprocess with `shell=true` as the shell to run cmd:
- local

These plugins use `C.DEFAULT_EXECUTABLE` as the shell to run cmd:
- chroot
- docker
- jail
- libvirt_lxc
- lxd

These plugins let the remote system select the shell to run cmd:
- funcd
- paramiko_ssh
- ssh
- zone

So it seems all the plugins (including this one) are about about as consistent as they can be, although the local plugin could be updated to not use `shell=true`.

Additionally, using `C.DEFAULT_EXECUTABLE` for the shell may not be the best choice, but if we're going to change that, we should do so for all of the plugins to keep things consistent.

bcoca commented:
it is a bit more subtle than  that, the reason we specify a shell on the action plugin/privilege escalation side is to control the quoting. 

The connection should not specify the shell as it does not have all the info that it needs to select the correct one, this happens in the command construction so there is no need of adding it again, the plugins that do so now need to be corrected.

mattclay commented:
Since `_low_level_execute_command` passes `cmd` to all connections as a string it's always being parsed by a shell specified by the connection.

The only difference between connections is how that shell is chosen:
- ssh - The shell of the ssh user on the remote host is used.
- lxd - The shell chosen by the connection plugin is used.

Either way the shell parsing `cmd` is selected without any awareness of quoting requirements for the command.

That's why `_low_level_execute_command` wraps `cmd` in an additional shell before passing it to `self._connection.exec_command`.

If there are scenarios where this doesn't work, changing some plugins to use a list isn't going to help, because others, like ssh, don't have that option, they can only pass a string to the remote host.

Is there a specific scenario where the current behavior with ssh breaks?

bcoca commented:
no, as ssh does not set it's own shell or quoting, it works in all scenarios, the ones that will break are the ones constructing the cmd with their own shell, like local and lxd

mattclay commented:
Correct, the ssh connection plugin does not set its own shell or quoting, but the remote host does.

Here's a demo showing the ssh and lxd connection plugins behave the same:

Give "tester" a custom shell so we can see the args it receives:

```
$ grep tester /etc/passwd
tester:x:1001:1001:,,,:/home/tester:/home/tester/shell
```

Here's the custom shell:

```
$ cat /home/tester/shell
#!/usr/bin/python
import sys
print('remote host shell: {0}'.format(sys.argv))
```

Here's a manual connection using ssh:

```
$ ssh tester@localhost 'command goes here'
remote host shell: ['/home/tester/shell', '-c', 'command goes here']
```

Here's a manual connection using lxd:

```
$ lxc exec centos-7-amd64 -- /home/tester/shell '-c' 'command goes here'
remote host shell: ['/home/tester/shell', '-c', 'command goes here']
```

Here's a connection using ansible ssh:

```
$ ansible -m ping -i 'localhost,' -e 'ansible_user=tester ansible_connection=ssh ansible_ssh_pipelining=true' all
localhost | FAILED! => {
    "changed": false, 
    "failed": true, 
    "module_stderr": "", 
    "module_stdout": "remote host shell: ['/home/tester/shell', '-c', \"/bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python'\"]\n", 
    "msg": "MODULE FAILURE", 
    "parsed": false
}
```

Here's a connection using ansible lxd (with '/home/tester/shell' instead of DEFAULT_EXECUTABLE):

```
$ ansible -m ping -i 'centos-7-amd64,' -e 'ansible_connection=lxd ansible_ssh_pipelining=true' all
centos-7-amd64 | FAILED! => {
    "changed": false, 
    "failed": true, 
    "module_stderr": "", 
    "module_stdout": "remote host shell: ['/home/tester/shell', '-c', \"/bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python'\"]\n", 
    "msg": "MODULE FAILURE", 
    "parsed": false
}
```

There is no difference in how ssh or lxd execute `cmd`. They both use a shell chosen by the connection to execute `cmd`.

What am I missing?

mattclay commented:
@bcoca: Once PR #14845 is merged I'll update this PR to use the new cmd list. I already have the changes locally, since I tested PR #14845 against this plugin.

mattclay commented:
I've rebased and squashed the requested fixes back into the original commit. At this point I think I'm just waiting on PR #14845 so I can update the lxd plugin to use cmd as a list.

mattclay commented:
Rebased and switched from `C.DEFAULT_EXECUTABLE` to the new `self._play_context.executable`.

There's still a double shell, as with other plugins, but the lxd plugin no longer chooses the shell.

mattclay commented:
Rebased to resolve merge conflicts. I'll squash when finished.

mattclay commented:
@bcoca: Now that this plugin is using `self._play_context.executable`, is there anything else you'd like changed (aside from squashing commits), before this can be merged?

When https://github.com/ansible/ansible/pull/14845 is completed I can update this plugin for the new cmd list, along with the other plugins.
